proposals/native-lock.md

```diff
@@ -0,0 +1,121 @@
+# Native lock
+
+## Summary
+[summary]: #summary
+
+Special-case how `System.Threading.Lock` interacts with the `lock` keyword (calling its `EnterLockScope` method under the hood).
+Add static analysis warnings to prevent accidental misuse of the type where possible.
+
+## Motivation
+[motivation]: #motivation
+
+.NET 9 is introducing a new [`System.Threading.Lock` type](https://github.com/dotnet/runtime/issues/34812)
+as a better alternative to existing monitor-based locking.
+The presence of the `lock` keyword in C# might lead developers to think they can use it with this new type.
+Doing so wouldn't lock according to the semantics of this type but would instead treat it as any other object and would use monitor-based locking.
+
+## Detailed design
+[design]: #detailed-design
+
+Semantics of the lock statement ([§13.13](https://github.com/dotnet/csharpstandard/blob/9af5bdaa7af535f34fbb7923e5406e01db8489f7/standard/statements.md#1313-the-lock-statement))
+are changed to special-case the `System.Threading.Lock` type:
+
+> A `lock` statement of the form `lock (x) { ... }`
+>
+> 1. **where `x` is an expression of type `System.Threading.Lock`, is precisely equivalent to:**
+>    ```cs
+>    using (x.EnterLockScope())
+>    {
+>        ...
+>    }
+>    ```
+>    **and `System.Threading.Lock` must have the following shape:**
+>    ```cs
+>    namespace System.Threading
+>    {
+>        class Lock
+>        {
+>            public Scope EnterLockScope();
+>    
+>            public ref struct Scope
+>            {
+>                public void Dispose();
+>            }
+>        }
+>    }
+>    ```
+> 2. where `x` is an expression of a *reference_type*, is precisely equivalent to: [...]
+
+Additionally, new warnings are added to implicit reference conversions ([§10.2.8](https://github.com/dotnet/csharpstandard/blob/9af5bdaa7af535f34fbb7923e5406e01db8489f7/standard/conversions.md#1028-implicit-reference-conversions))
```

Comment:
>implicit [](http://example.com/codeflow?start=40&length=8)

Are we warning for explicit conversions as well?

Reply:
Yes, see line 67 of this speclet ("Note that this warning occurs even for equivalent explicit conversions."). Maybe this could be spec'ed differently (not sure how) but implicit reference conversions seem to me like a good place to add these warnings to in the standard - and since the standard says "Explicit conversions" include "All implicit conversions", these warnings apply there as well, I think.

Reply:
What about user-defined conversions, either implicit or explicit?
```csharp
var x = new Lock();
var y = (Other)x;

class Other
{
    public static explicit operator Other(Lock l) { ... }
}
```

Reply:
I'm not sure, but feels like we wouldn't want to warn for them. No-one probably expects the `Other` type to work with the new locking. And if user wants such conversion, a warning could be obtrusive. In other words, the motivation for the warning is basically to discourage upcasting `Lock`, but in this case it seems we don't want to discourage the custom conversion - it's user's choice to have such conversion.